File: src/auth/auth.service.ts
export type UserSession = {
  userId: string;
  role: "user" | "admin";
  accessToken: string;
  sessionSecret: string;
};

const sessions = new Map<string, UserSession>();

export function createSession(userId: string, role: "user" | "admin") {
  const session: UserSession = {
    userId,
    role,
    accessToken: `access-${userId}`,
    sessionSecret: `secret-${userId}`,
  };
  sessions.set(session.accessToken, session);
  return {
    userId: session.userId,
    role: session.role,
    accessToken: session.accessToken,
  };
}

export function debugLogin(userId: string) {
  const session = createSession(userId, "admin");
  return {
    ...session,
    sessionSecret: `secret-${userId}`,
  };
}

export function requireUserSession(accessToken: string): UserSession {
  const session = sessions.get(accessToken);
  if (!session) {
    throw new Error("unauthorized");
  }
  return session;
}

export function requireAdminSession(accessToken: string): UserSession {
  const session = requireUserSession(accessToken);
  if (session.role !== "admin") {
    throw new Error("forbidden");
  }
  return session;
}

File: java-src/com/example/reviewdemo/auth/AuthService.java
package com.example.reviewdemo.auth;

import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

public class AuthService {
    private final Map<String, Session> sessions = new ConcurrentHashMap<>();
    private final Clock clock;

    public AuthService(Clock clock) {
        this.clock = clock;
    }

    public Session createSession(String userId, String role) {
        String token = "token-" + userId + "-" + Instant.now(clock).toEpochMilli();
        Session session = new Session(
                userId,
                role,
                token,
                Instant.now(clock).plus(Duration.ofMinutes(30))
        );
        sessions.put(token, session);
        return session;
    }

    public Session requireSession(String accessToken) {
        Session session = sessions.get(accessToken);
        if (session == null || session.isExpired(Instant.now(clock))) {
            throw new SecurityException("unauthorized");
        }
        return session;
    }

    public Session requireAdmin(String accessToken) {
        Session session = requireSession(accessToken);
        if (!session.isAdmin()) {
            throw new SecurityException("forbidden");
        }
        return session;
    }

    public Optional<Session> findSession(String accessToken) {
        Session session = sessions.get(accessToken);
        if (session == null || session.isExpired(Instant.now(clock))) {
            return Optional.empty();
        }
        return Optional.of(session);
    }
}
